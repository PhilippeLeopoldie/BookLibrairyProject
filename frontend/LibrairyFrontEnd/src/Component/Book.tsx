import React, {useEffect, useState} from 'react'
 import {BookType} from '../Type'
 import FetchApi from '../FetchApi'
import PostOpinion from './PostOpinion'
 

 
function Books () {
    const [books, setBooks] = useState<BookType[]>()
    useEffect(() =>{
        FetchApi('http://localhost:5133/api/Books').then(books => setBooks(books))
    },[])
 //console.log(books?.at(Math.floor(Math.random()*books.length)))
 let random = books?.at(Math.floor(Math.random()*books.length))

    return (
        <>
        <h2>Suggestion of the day</h2>
        <h3>{random?.title}</h3>
        {
             books?.map((book, index)=>(
                <p key={index}>{book.title}{book.author}
                <PostOpinion bookId = {book.bookId}/></p>
                   
            )) 
            
        }
    </>
    )
}
 export default Books
