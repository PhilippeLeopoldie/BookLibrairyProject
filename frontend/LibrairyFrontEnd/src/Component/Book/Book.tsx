import { Opinion } from "../Opinion/Opinion";
import url from "../../Url";
import trash from "../../media/delete.svg";
import { useState } from "react";
import { OpinionEdit } from "../Opinion/OpinionEdit";

type BookWithDeletionHandler = {
  book: {
    id: number;
    title: string;
    author: string;
  };
  onDelete: Function;
};

export const Book = ({ book, onDelete }: BookWithDeletionHandler) => {
  const [editHandling, setEditHandling] = useState<Boolean>(false);

  const toggleEditOpinion = () => {
    setEditHandling(!editHandling);
  };

  const DeleteBook = async (bookId: number) => {
    try {
      await fetch(url + `api/Book/${bookId}`, {
        method: "DELETE",
      });
      onDelete(bookId);
    } catch (error) {
      console.error("Error deleting book:", error);
    }
  };

  return (
    <>
      {!editHandling ? (
        <div className="bookcard" key={book.id}>
          <div className="bookcard--header">
            <button
              className=" button booktitle--trashbutton"
              type="submit"
              onClick={async () => {
                await DeleteBook(book.id);
              }}
            >
              <img
                className="icone bookcard--iconeTrash"
                src={trash}
                alt="delete"
              />
            </button>
            <h2 className="booktitle">{book.title}</h2>
            <h3 className="bookauthor">by: {book.author}</h3>
          </div>
          <Opinion bookId={book.id} onEdit={toggleEditOpinion} />
        </div>
      ) : (
        <OpinionEdit onEdit={toggleEditOpinion} />
      )}
    </>
  );
};
