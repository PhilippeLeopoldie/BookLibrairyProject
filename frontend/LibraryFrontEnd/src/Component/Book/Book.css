.booksContainer {
  gap: 2%;
  display: grid;
  justify-content: center;
}

.bookcard--grid {
  border: 1px gray solid;
  border-radius: 3%;
  display: grid;
  grid-template-rows: 2fr 3fr 1fr;
  width: 20rem;
  height: 30rem;
  border-radius: 20px 16px 12px 32px;
  background-image: linear-gradient(
    to right,
    #d4af37 10%,
    rgb(253, 229, 187) 10%
  );
}
.bookCard__RateButton{
 margin-top :4%;
 margin-bottom: 8%;
 margin-left: 30%;
 margin-right: 20%;
 font-size:large;
 font-weight:bolder;
  color:saddlebrown; 
 background-color: transparent;
 border-color: #a38729;

}

.bookcard--grid:after {
  display: flex;
  justify-content: end;
  margin-left: 2%;
  height: 52px;
  bottom: 6px;
  right: 0px;
  background: white;
  border-radius: 32px 4px 4px 32px;
  box-shadow: inset 4px 6px 0px 0px #a69650;
  background-image: linear-gradient(
    to bottom,
    transparent 6px,
    #a69650 8px,
    transparent 8px,
    transparent 12px,
    #a69650 12px,
    transparent 14px,
    transparent 18px,
    #a69650 18px,
    transparent 20px,
    transparent 24px,
    #a69650 24px,
    transparent 26px,
    transparent 30px,
    #a69650 30px,
    transparent 32px,
    transparent 36px,
    #a69650 36px,
    transparent 38px,
    transparent 42px,
    #a69650 42px,
    transparent 44px,
    transparent 48px,
    #a69650 48px,
    transparent 50px
  );
  content: "";
  position: relative;
}

.bookcard--header {
  border-top-right-radius: 5%;
  border-top-left-radius: 5%;
  background-color: transparent;
  padding-left: 10%;
}

.booktitle {
  background-color: transparent;
  font-size: x-large;
  font-weight: bold;
  -webkit-text-stroke: 1px #a38729;
  color: #d4af37;
  margin: 0;
}

.bookauthor {
  background-color: transparent;
  margin:2%;
  margin-bottom: 8%;
}

.bookform--view{
  resize: none;
}

.booktitle--trashbutton {
  display: flex;
  width: 40px;
  height: 40px;
  margin: 2%;
}

.bookcard--iconeTrash {
  border-radius: 20%;
}

@media screen and (min-width: 395px) {
  .bookcard {
    width: 20rem;
  }
}

@media screen and (min-width: 698px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr;
  }
}

@media screen and (min-width: 1042px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

@media screen and (min-width: 1458px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr 1fr 1fr;
  }
}
