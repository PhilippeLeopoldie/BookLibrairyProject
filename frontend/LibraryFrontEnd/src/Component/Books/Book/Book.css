.bookcard--grid {
  border: 1px gray solid;
  display: grid;
  width: 18rem;
  height: 29rem;
  margin: 0 1rem 2rem 1rem;
  background-color: #ffffff;
  border-radius: 10px;
}

.bookcard--black {
  background-color: #000000;
  border: 1px #ffffff solid;
  .booktitle,
  .bookauthor {
    color: #ffffff;
  }
}


.boocard__Image {
  border-radius: 9px 9px 0 0;
  border-bottom: gray 1px solid;
  cursor: pointer;
}

.boocard__Image-zoomIn {
  transform: scale(1.2);
  border:gray 1px solid;
}

.boocard__Image-zoomOut {
  transform: scale(1);
}

.bookcard__footer--flex {
  display: flex;
  max-width: 18rem;
  align-items: center;
  justify-content: space-between;
  background-color: transparent;
}

.footer__average-rate--flex{
  display:flex;
  background-color: transparent;
}

.footer__average-rate {
  font-weight: bold;
  text-align: right;
  color: rgb(0, 76, 255);
  text-decoration: underline;
  cursor: pointer;
  background-color: transparent;
  margin-left: 10%;
}

.rate_star {
  margin-top: -1rem;
}

.bookCard__RateButton {
  margin-bottom: 4%;
  margin-right: 3%;
  width: 50%;
  font-size: 1rem;
  font-weight: bolder;
  color: rgb(0, 76, 255);
}

.bookcard__header {
  border-top-right-radius: 5%;
  border-top-left-radius: 5%;
  background-color: transparent;
}

.booktitle {
  background-color: transparent;
  font-size: 1.4rem;
  font-weight: bold;
  text-align: center;
  margin-top: 0;
  margin-bottom: 0.5rem;
  border-color: transparent;
  resize: none;
  max-height: 3.3rem;
  width: 94%;
  border-top-right-radius: 10%;
}

footer {
  background-color: transparent;
}

.bookauthor {
  background-color: transparent;
  margin-top: 0.2rem;
  margin-bottom: 0.2rem;
}

.bookform--view {
  resize: none;
}

.booktitle--trashbutton {
  display: flex;
  width: 25px;
  height: 30px;
  margin: 2%;
  padding-left: 0;
}

.bookcard--iconeTrash {
  border-radius: 20%;
  height: 25px;
}
