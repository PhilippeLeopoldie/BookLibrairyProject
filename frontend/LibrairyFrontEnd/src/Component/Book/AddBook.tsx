import { useEffect, useState } from "react";
import { BookType } from "../../Type";
import FetchApi from "../../FetchApi";
import url from "../../Url";
import love from "../../media/love.png";
import sad from "../../media/sad.png";
import "./Book.css";

function AddBook() {
  const [formData, setFormData] = useState({
    title: "",
    author: "",
    view: "",
    userName: "",
    rate:""
  });
  const [books, setBooks] = useState<BookType[]>([]);
  const [bookCreatedMessage, setBookCreatedMessage] = useState<boolean>(false);

  const PostBook = async () => {
    const requestOptions = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title: formData.title,
        author: formData.author,
      }),
    };
    const response = await fetch(url + "api/Book", requestOptions);
    const body = await response.json();
    console.log("postBook bodyId:", body.id);
    setBooks((books) => [...books, body]);
    return body;
  };
  const fetchBooks = async () => {
    try {
      const booksData = await FetchApi(url + "api/Book");
      setBooks(booksData.$values);
    } catch (error) {
      console.error("error fetching books:", error);
    }
  };

  useEffect(() => {
    fetchBooks();
  }, []);

  const PostOpinion = async (bookId: number) => {
    const requestOptions = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        bookId,
        rate:formData.rate,
        view: formData.view,
        userName: formData.userName,
      }),
    };
    const OpinionResponse = await fetch(url + "api/Opinion", requestOptions);
    const newOpinion = await OpinionResponse.json();

    // clear the form fields after submitting
    setFormData({
      title: "",
      author: "",
      view: "",
      userName: "",
      rate:""
    });
    setBookCreatedMessage(true);
    return newOpinion;
  };
  const HideBookCreatedMessage = () => {
    setBookCreatedMessage(false);
  };
  const HandleInputChange = (
    e:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    setFormData({ ...formData, [e.target.name]: e.target.value });
    bookCreatedMessage && HideBookCreatedMessage();
  };

  return (
    <>
      <form
        onSubmit={(e) => e.preventDefault()}
        action={url + "api/Book"}
        method="POST"
        className="bookform"
      >
        <div className="bookform__inputs">
          <input
            className="input"
            placeholder="Title"
            name="title"
            value={formData.title}
            onChange={(e) => HandleInputChange(e)}
          />

          <input
            className="input"
            placeholder="Author"
            name="author"
            value={formData.author}
            onChange={(e) => HandleInputChange(e)}
          />

          <textarea
            className="bookform__view input"
            placeholder="View"
            name="view"
            value={formData.view}
            onChange={(e) => HandleInputChange(e)}
          />

          <input
            className="input"
            placeholder="UserName"
            name="userName"
            value={formData.userName}
            onChange={(e) => HandleInputChange(e)}
          />

          <input
            type="number"
            className="input"
            placeholder ="Rate"
            name="rate"
            value={formData.rate}
            onChange={(e) => HandleInputChange(e)}
          />
        </div>

        <div>
          <button
            className="button bookform__button"
            onClick={async () => {
              PostOpinion((await PostBook()).id);
            }}
          >
            <img className="icone iconeRate" src={sad} />
            Dislike
          </button>

          <button
            className="button bookform__button"
            type="submit"
            onClick={async () => {
              PostOpinion((await PostBook()).id);
            }}
          >
            <img className="icone iconeRate" src={love} />
            Like
          </button>
        </div>
        <div className="bookform__output">
          {bookCreatedMessage && <h1 className="bookform__output">Book created!</h1>}
        </div>
      </form>
      {/* <Books /> */}
    </>
  );
}

export default AddBook;
