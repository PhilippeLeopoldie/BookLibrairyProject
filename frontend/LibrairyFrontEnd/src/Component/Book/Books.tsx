import { useEffect, useState } from "react";
import { BookType} from "../../Type";
import FetchApi from "../../FetchApi";
import url from "../../Url";
import {Book} from "./Book";


export const Books = () => {
  const [books, setBooks] = useState<BookType[]>([]);
  const random = books?.at(Math.floor(Math.random() * books.length));

  const handleDelete = (bookId : number) => {
    setBooks((books) => books.filter((book) => book.book.id !== bookId))
  }
  const fetchBooks = async () => {
    try {
      const booksData = await FetchApi(url + "api/Book");
      setBooks(booksData.$values);  
    } catch (error) {
      console.error("Error fetching books:", error);
    }
  };

  useEffect(() => {
    fetchBooks();
  }, []);

  return (
    <div className="books">
      <h2>Recommendation of the day</h2>
      <h3>{random?.book.title}</h3>
      <div className="bookcontainer">
        {books &&
          books
          .sort((a,b) => b.book.id - a.book.id)
          .map((bookDetail, index) => (
            <Book book = {bookDetail.book} onDelete ={handleDelete}/>
          ))}
      </div>
    </div>
  );
}
