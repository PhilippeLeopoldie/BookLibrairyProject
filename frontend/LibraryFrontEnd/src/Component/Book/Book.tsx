import "./Book.css";
import { Opinion } from "../Opinion/Opinion";
import { OpinionCreate } from "../OpinionCreate/OpinionCreate";
import { OpinionList } from "../OpinionList/OpinionList";
import { ThemeContext } from "../App/App";
import trash from "../../media/delete.svg";
import url from "../../Url";
import { useContext, useState } from "react";

type BookWithDeletionHandler = {
  book?: {
    id: number;
    title: string;
    author: string;
    imageUrl?:string;
  };
  onDelete?: Function;
};

type Reviews = {
  id: number;
  rate: number;
  view: string;
  userName: string;
  postDate: string;
  bookId: number;
};

export const Book = ({ book, onDelete }: BookWithDeletionHandler) => {
  const theme = useContext(ThemeContext);
  const [createOpinionHandling, setCreateOpinionHandling] =
    useState<boolean>(false);
  const [displayReview, setDisplayReview] = useState<Boolean>(false);
  const [reviewList, setReviewList] = useState<Reviews[]>();

  const toggleOpinionList = () => {
    setDisplayReview(!displayReview);
  };
  const handleOpinionList = (reviews: Reviews[]) => {
    toggleOpinionList();
    setReviewList(reviews);
  };

  const toggleCreateOpinion = () => {
    setCreateOpinionHandling(!createOpinionHandling);
  };

  const DeleteBook = async (bookId: number) => {
    try {
      await fetch(url + `api/Book/${bookId}`, {
        method: "DELETE",
      });
      onDelete && onDelete(bookId);
    } catch (error) {
      console.error("Error deleting book:", error);
    }
  };

  return (
    <>
      {displayReview ? (
        <OpinionList
          opinions={reviewList}
          displayReviews={() => toggleOpinionList()}
          book={book}
        />
      ) : book && !createOpinionHandling ? (
        <div className={"bookcard--grid bookcard--" + theme}>
          <img
            src={book.imageUrl &&`${book.imageUrl}`}
            className="boocard__Image"
            alt="bookImage"
            width="288px"
            height="326px"
          />
          <header className="bookcard__header">
            {/* <button
            
              className=" button booktitle--trashbutton"
              type="submit"
              onClick={async () => {
                await DeleteBook(book.id);
              }}
            >
              <img
                className="icone bookcard--iconeTrash"
                src={trash}
                alt="delete"
              />
            </button> */}
            <h3  
              title="Book Title"
              className="booktitle">{book.title.length > 40 
                ? `${book.title.slice(0, 40)}...`
                : book.title}
            </h3>
            <h3 className="bookauthor">by: {book.author}</h3>
          </header>
          <footer className="bookcard__footer--flex">
          <Opinion
            book={book}
            toCreate={toggleCreateOpinion}
            displayReview={handleOpinionList}
          />
          
            <button
              className="button bookCard__RateButton"
              onClick={() => {
                toggleCreateOpinion();
              }}
            >
              Rate this book
            </button>
          </footer>
        </div>
      ) : (
        <OpinionCreate book={book && book} toCreate={toggleCreateOpinion} />
      )}
    </>
  );
};
