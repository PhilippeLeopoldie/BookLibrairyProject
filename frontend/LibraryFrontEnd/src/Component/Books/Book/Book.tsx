import "./Book.css";
import { Opinion } from "../../Opinions/Opinion/Opinion";
import { OpinionCreate } from "../../Opinions/OpinionCreate/OpinionCreate";
import { OpinionList } from "../../Opinions/OpinionList/OpinionList";
import { ThemeContext } from "../../App/App";
import { BOOK_BY_BOOKID_URL } from "../../../Url";
import { useContext, useEffect, useState } from "react";

type BookType = {
  book: {
    id: number;
    title: string;
    author: string;
    averageRate: number;
    imageUrl?: string;
  };
};

type Reviews = {
  id: number;
  rate: number;
  view: string;
  userName: string;
  postDate: string;
  bookId: number;
};

export const Book = ({ book }: BookType) => {
  const theme = useContext(ThemeContext);
  const [updatedBook, setUpdatedBook] = useState<BookType>({ book });
  const [displayCreateOpinion, setDisplayCreateOpinion] =
    useState<boolean>(false);
  const [opinionCreated, setOpinionCreated] = useState<boolean>(false);
  const [displayReview, setDisplayReview] = useState<Boolean>(false);
  const [averageClick, setAverageclick] = useState<Boolean>(false);
  const [reviewList, setReviewList] = useState<Reviews[]>();

  const fetchBook = async () => {
    try {
      const bookResponse: Response = await fetch(
        `${BOOK_BY_BOOKID_URL}${book?.id}`
      );
      if (bookResponse.status === 200) {
        const bookResponseData = await bookResponse.json();
        setUpdatedBook(bookResponseData);
      } else if (bookResponse.status === 404) {
        console.log(bookResponse);
      }
    } catch (error) {
      console.error(`Error fetching bookId:${book?.id}`, error);
    }
  };
  useEffect(() => {
    opinionCreated === true && fetchBook();
    console.log("opinionCreated status: ", opinionCreated);
  }, [opinionCreated]);

  const toggleOpinionList = () => {
    setDisplayReview(!displayReview);
    setAverageclick(false);
  };
  const handleOpinionList = (reviews: Reviews[]) => {
    toggleOpinionList();
    setReviewList(reviews);
  };

  const toggleCreateOpinion = () => {
    setDisplayCreateOpinion(!displayCreateOpinion);
  };

  const handleCreatedOpinion = (value: boolean) => {
    setOpinionCreated(value);
  };

  return (
    <>
      {displayReview ? (
        <OpinionList
          opinions={reviewList}
          displayReviews={() => toggleOpinionList()}
          book={book}
        />
      ) : book && !displayCreateOpinion ? (
        <div className={"bookcard--grid bookcard--" + theme}>
          <img
            src={book.imageUrl && `${book.imageUrl}`}
            className="boocard__Image"
            alt="bookImage"
            width="288px"
            height="326px"
          />
          <header className="bookcard__header">
            <h3 title="Book Title" className="booktitle">
              {book.title.length > 40
                ? `${book.title.slice(0, 40)}...`
                : book.title}
            </h3>
            <h3 className="bookauthor">by: {book.author}</h3>
          </header>
          <footer className="bookcard__footer--flex">
            <div className="footer__average-rate--flex">
              <a
                className="footer__average-rate"
                onClick={() => {
                  console.log("averageClick=", averageClick);
                  setAverageclick(!averageClick);
                }}
              >
                {book.averageRate}/5
              </a>
              <div className="rate_star"> &#9733;</div>
            </div>
            {averageClick && (
              <Opinion
                book={updatedBook?.book}
                displayReview={handleOpinionList}
              />
            )}
            
            <button
              className="button bookCard__RateButton"
              onClick={() => {
                toggleCreateOpinion();
              }}
            >
              Rate this book
            </button>
          </footer>
          
        </div>
      ) : (
        <OpinionCreate
          book={book && book}
          toCreate={toggleCreateOpinion}
          created={handleCreatedOpinion}
        />
      )}
    </>
  );
};
