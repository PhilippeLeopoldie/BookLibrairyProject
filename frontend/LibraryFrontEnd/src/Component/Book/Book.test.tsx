import {render, screen,} from "@testing-library/react";
import { Book } from "./Book";
import "@testing-library/jest-dom";

const actualBookData = {
  id: 1,
  title: "The Great Gatsby",
  author: "F. Scott Fitzgerald",
};

test("renders book title and author", () => {
  // Arrange
  render(<Book book={actualBookData} onDelete={() => {}} />);

  // Act
  const titleElement = screen.getByText("The Great Gatsby");
  const authorElement = screen.getByText("by: F. Scott Fitzgerald");

  // Assert
  expect(titleElement).toBeInTheDocument();
  expect(authorElement).toBeInTheDocument();
  expect(titleElement).toHaveClass("booktitle");
  expect(authorElement).toHaveClass("bookauthor");
});

test("renders delete img with the correct alt text and ClassName", () => {
  // Arrange
  const onDelete = jest.fn();
  const { getByAltText } = render(
    <Book book={actualBookData} onDelete={onDelete} />
  );

  // Act
  const trashImage = getByAltText("delete");

  //Assert
  expect(trashImage).toBeInTheDocument();
  expect(trashImage).toHaveClass("bookcard--iconeTrash");
});
