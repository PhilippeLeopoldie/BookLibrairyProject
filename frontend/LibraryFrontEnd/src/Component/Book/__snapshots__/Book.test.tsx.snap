// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` snapchot: renders book title and author 1`] = `
<DocumentFragment>
  <div
    class="bookcard--grid"
  >
    <header
      class="bookcard--header"
    >
      <button
        class=" button booktitle--trashbutton"
        type="submit"
      >
        <img
          alt="delete"
          class="icone bookcard--iconeTrash"
          src="SvgrURL"
        />
      </button>
      <textarea
        class="booktitle "
        readonly=""
      >
        The Great Gatsby
      </textarea>
      <h3
        class="bookauthor"
      >
        by: F. Scott Fitzgerald
      </h3>
    </header>
    <h2
      class="OpinionLoading"
    >
      Loading...
    </h2>
    <footer>
      <button
        class="button bookCard__RateButton"
      >
        Rate this book
      </button>
    </footer>
  </div>
</DocumentFragment>
`;
